### PR TITLE
use array interface to lib*.d functions

### PR DESCRIPTION
**Pull request by WalterBright**
Also only interface to C++ what is actually needed.

### CODE REVIEW

File: src/lib.h [hunk 1]
```diff
@@ -9,39 +9,5 @@
...
-    virtual void setFilename(const char *dir, const char *filename) = 0;
-    virtual void addObject(const char *module_name, void *buf, size_t buflen) = 0;
-    virtual void addLibrary(void *buf, size_t buflen) = 0;
-    virtual void write() = 0;
-};
-
-#endif /* DMD_LIB_H */
-
+class Library;
+void addObjectToLibrary(Library *lib, const char *module_name, const unsigned char *buf, size_t buflen);
```
Comment (yebblies) [inline src/lib.h]: This should just go in the top of glue.c, we don't need this header.

Comment (WalterBright) [inline src/lib.h]: you're right

**Comment by ZombineDev** [comment]
LGTM

You can also change `Library.addLibrary` to:
```
void addLibrary(const ubyte[] buf)
{
       addObject(null, buf);
 }
```
and delete the overrides in `LibElf`, `LibMach`, `LibMSCoff` and `LibOMF` (unless there's a good reason for keeping the method `abstract` that I'm missing), but I'm fine with leaving that for another PR.

**Comment by WalterBright** [comment]
> I'm fine with leaving that for another PR.

Thank you. It's awfully tempting to keep adding one more thing, but I'm a big proponent of keeping PRs small, focused, and incremental.

**Comment by yebblies** [comment]
Auto-merge toggled on

**Comment by MartinNowak** [comment]
Breaks OSX-32 build, apparently due to the size_t parameter. Can we please finally fix this? There must be some mangling in OSX-32 for what we call size_t in D.

**Comment by MartinNowak** [comment]
Made a ticket [Issue 16289 – no extern(C++) mangling for size_t/unsigned long parameters](https://issues.dlang.org/show_bug.cgi?id=16289).

**Comment by yebblies** [comment]
Why are we building OSX32?  It is not a supported host platform, that's why the autotester isn't testing it.

**Comment by MartinNowak** [comment]
> Why are we building OSX32? It is not a supported host platform, that's why the autotester isn't testing it.

That's what you already asked last time, and you also got an answer, see https://github.com/dlang/dmd/pull/5788#issuecomment-219967689. As other people still rely on interoperability w/ OSX-32, we shouldn't drop it over such a simple issue. Also the issue is more general, see the Bugzilla link.

**Comment by yebblies** [comment]
And nothing has changed, it's still a bad idea.  Time to fix the build script to not build for this dead platform.  We've already dropped it, the decision was made years ago.

**Comment by MartinNowak** [comment]
The C++ size_t mangling issue is a problem w/ OSX-32 targets which we do support.

**Comment by JackStouffer** [comment]
> OSX-32 targets which we do support

Might as well be supporting Windows XP for all the good it will do you.

**Comment by yebblies** [comment]
> The C++ size_t mangling issue is a problem w/ OSX-32 targets which we do support.

Has nothing to do with supporting OSX-32 as a host platform.  Why don't we just stop testing it and let it die?

**Pull request merged**